### PR TITLE
tests linearizability: reproduce and prevent 14571

### PR DESCRIPTION
**Pull request by chaochn47**
Related to https://github.com/etcd-io/etcd/issues/14045

Signed-off-by: Chao Chen <chaochn@amazon.com>

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Review by ptabor — APPROVED** [review]

**Review by ptabor — CHANGES_REQUESTED** [review]
(I'm sorry - prematurly clicked approve)

File: tests/linearizability/traffic.go [hunk 1]
```diff
@@ -105,3 +109,37 @@ func (t readWriteSingleKey) pickWriteOperation() Operation {
 	}
 	panic("unexpected")
 }
+
+func (t readWriteSingleKey) AuthEnabled() bool { return false }
+
+type authTraffic struct {
```
Comment (serathius) [inline tests/linearizability/traffic.go:115]: Can auth be just a bool flag on `readWriteSingleKey`?

File: tests/linearizability/linearizability_test.go [hunk 2]
```diff
@@ -43,77 +45,112 @@ const (
...
 }
 
-func testLinearizability(ctx context.Context, t *testing.T, config e2e.EtcdProcessClusterConfig, failpoint FailpointConfig, traffic trafficConfig) {
+func testLinearizability(ctx context.Context, t *testing.T, config e2e.EtcdProcessClusterConfig, failpoint FailpointConfig, traffic trafficConfig, lg *zap.Logger) {
 	clus, err := e2e.NewEtcdProcessCluster(ctx, t, e2e.WithConfig(&config))
 	if err != nil {
 		t.Fatal(err)
 	}
 	defer clus.Close()
+	if traffic.traffic.AuthEnabled() {
```
Comment (serathius) [inline tests/linearizability/linearizability_test.go:131]: Setting up authority for traffic should be done by traffic.

File: tests/linearizability/linearizability_test.go [hunk 3]
```diff
@@ -43,77 +45,112 @@ const (
...
+			name: "Issue14571",
+			failpoint: FailpointConfig{
+				failpoint: MemberAddFailpoint,
+				count:     1,
+			},
+			config: e2e.EtcdProcessClusterConfig{
+				ClusterSize:   1,
+				SnapshotCount: 2,
+			},
+			clientCount: 3,
```
Comment (serathius) [inline tests/linearizability/linearizability_test.go:101]: Why this requires 3 clients?

File: tests/linearizability/auth.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+const (
+	rootUserName     = "root"
+	rootRoleName     = "root"
+	rootUserPassword = "123"
+	testUserName     = "test-user"
+	testRoleName     = "test-role"
+	testUserPassword = "abc"
+)
+
+func setupAuth(ctx context.Context, t *testing.T, c interfaces.Client) {
```
Comment (serathius) [inline tests/linearizability/auth.go:35]: Please move this to traffic

File: tests/linearizability/traffic.go [hunk 5]
```diff
@@ -18,17 +18,21 @@ import (
...
 
 var (
 	DefaultTraffic Traffic = readWriteSingleKey{key: "key", writes: []opChance{{operation: Put, chance: 100}}}
+	AuthTraffic    Traffic = authTraffic{readWriteSingleKey{key: "key", writes: []opChance{{operation: Put, chance: 100}}}}
 )
 
 type Traffic interface {
-	Run(ctx context.Context, c *recordingClient, limiter *rate.Limiter)
+	Run(ctx context.Context, c *recordingClient, limiter *rate.Limiter, lg *zap.Logger)
+	AuthEnabled() bool
```
Comment (serathius) [inline tests/linearizability/traffic.go:35]: I think traffic can handle it's own setup. Maybe instead of exposing `AuthEnabled() bool` we have `Setup(ctx context.Context, c interfaces.Client, lg *zap.Logger)`

File: tests/linearizability/failpoints.go [hunk 6]
```diff
@@ -228,3 +230,23 @@ func (f randomFailpoint) Trigger(t *testing.T, ctx context.Context, clus *e2e.Et
...
+
+func (f memberAddFailpoint) Trigger(t *testing.T, ctx context.Context, clus *e2e.EtcdProcessCluster) error {
+	// write more than SnapshotCount keys to single leader to make sure snapshot is created
+	testUserClientOpts := e2e.WithAuth("test-user", "abc")
+	for i := 0; i <= 10; i++ {
+		if err := clus.Client(testUserClientOpts).Put(ctx, "key", "test", config.PutOptions{}); err != nil {
+			panic(err)
+		}
+	}
+	if err := clus.StartNewProc(ctx, nil, t, e2e.WithAuth("root", "123")); err != nil {
```
Comment (serathius) [inline tests/linearizability/failpoints.go:244]: Not sure how running a new member is a trigger. Could we instead setup 3 member cluster, terminate one of the members (could be gradefully) and wait until there are enough transaction to trigger snapshot before starting it again?

File: tests/linearizability/linearizability_test.go (hunk 2 above)
Comment (chaochn47) [reply tests/linearizability/linearizability_test.go:131]: The initial authority is set up once ideally. 

If moving to `simulateTraffic`, it will be ran by multiple clients.

File: tests/linearizability/linearizability_test.go (hunk 3 above)
Comment (chaochn47) [reply tests/linearizability/linearizability_test.go:101]: Only 3 clients are needed to reproduce on v3.5.5.

File: tests/linearizability/auth.go (hunk 4 above)
Comment (chaochn47) [reply tests/linearizability/auth.go:35]: Looks like the same comment about moving the auth set up into traffic.

File: tests/linearizability/failpoints.go (hunk 6 above)
Comment (chaochn47) [reply tests/linearizability/failpoints.go:244]: Yeah, I can try this out.

File: tests/linearizability/traffic.go (hunk 5 above)
Comment (chaochn47) [reply tests/linearizability/traffic.go:35]: Okay. I can add one auth related configuration to handle the set up.

**Comment by chaochn47** [comment]
@serathius

> Not sure how running a new member is a trigger. Could we instead setup 3 member cluster, terminate one of the members (could be gradefully) and wait until there are enough transaction to trigger snapshot before starting it again?

I am not able to reproduce the issue with this set up. 

Would you mind shed some lights on if I set up the framework incorrectly? Here is the diff in my fork. https://github.com/chaochn47/etcd/pull/7

image

**Comment by chaochn47** [comment]
I think I get it reproduced with `100` compared to original `5000` `DefaultSnapshotCatchUpEntries`. With the original value, porcupine checking all the operations takes too long and test is killed after 10mins. 

Will try to not to include those recorded operations before the "real" failure is triggered. 

Does it make sense before I proceed with this approach?

File: tests/linearizability/failpoints.go [hunk 7]
```diff
@@ -99,13 +103,92 @@ func (f killFailpoint) Trigger(t *testing.T, ctx context.Context, clus *e2e.Etcd
...
+	initialRevision, err := getRevision(ctx, cc)
 	if err != nil {
 		return err
 	}
+	lg.Info("got revision",
+		zap.Int64("initial-revision", initialRevision),
+		zap.Int64("snapshot-catchup-count", snapshotCount),
+	)
+	for {
+		time.Sleep(time.Second)
```
Comment (lavacat) [inline tests/linearizability/failpoints.go:163]: nit: this doesn't abide 1s triggerTimeout.

File: tests/linearizability/failpoints.go [hunk 8]
```diff
@@ -99,13 +103,92 @@ func (f killFailpoint) Trigger(t *testing.T, ctx context.Context, clus *e2e.Etcd
...
+			DialKeepAliveTimeout: 5 * time.Millisecond,
+			Username:             rootUserName,
+			Password:             rootUserPassword,
+		})
+		require.NoError(t, err)
+		require.NoError(t, addTestUserAuth(ctx, cc))
+		// 5000 is the default number of snapshot count catch up entries
+		// DefaultSnapshotCatchUpEntries = 5000
+		// Try manipulating e2e etcd test binary to 100 locally and snapshot-count here to speed up test
+		require.NoError(t, waitForSnapshot(ctx, cc, 5000, lg))
```
Comment (lavacat) [inline tests/linearizability/failpoints.go:130]: are you trying to hit `s.authStore.Recover` etcdserver/server.go:1042? If yes, I think you need to bring back killed member, so it will actually apply snapshot.  Also you need to set snapshot-count to lower number. See https://github.com/etcd-io/etcd/issues/10219 for an example.

FYI, seams we are both working on reproducing `apply snapshot` path. My logic for https://github.com/etcd-io/etcd/issues/14726 looks similar, except I use goPanicFailpoint.

File: tests/linearizability/failpoints.go [hunk 9]
```diff
@@ -77,13 +79,15 @@ var (
...
 }
 
-type killFailpoint struct{}
+type killFailpoint struct {
+	waitForSnapshot bool
+}
 
-func (f killFailpoint) Trigger(t *testing.T, ctx context.Context, clus *e2e.EtcdProcessCluster) error {
+func (f killFailpoint) Trigger(t *testing.T, ctx context.Context, clus *e2e.EtcdProcessCluster, lg *zap.Logger) error {
 	member := clus.Procs[rand.Int()%len(clus.Procs)]
```
Comment (lavacat) [inline tests/linearizability/failpoints.go:91]: how about using `member := clus.Procs[(clus.WaitLeader(t)+1)%len(clus.Procs)]` to skip leader election and save time? Or do you want to test killing leader?

Comment (chaochn47) [reply tests/linearizability/failpoints.go:91]: The kill failpoint should not only applicable to followers. 2s wait time is fine.

File: tests/linearizability/failpoints.go [hunk 10]
```diff
@@ -99,13 +103,92 @@ func (f killFailpoint) Trigger(t *testing.T, ctx context.Context, clus *e2e.Etcd
...
+			Logger:               zap.NewNop(),
+			DialKeepAliveTime:    1 * time.Millisecond,
+			DialKeepAliveTimeout: 5 * time.Millisecond,
+			Username:             rootUserName,
+			Password:             rootUserPassword,
+		})
+		require.NoError(t, err)
+		require.NoError(t, addTestUserAuth(ctx, cc))
+		// 5000 is the default number of snapshot count catch up entries
+		// DefaultSnapshotCatchUpEntries = 5000
```
[the remark this answers is not in the document]
Comment (chaochn47) [reply tests/linearizability/failpoints.go:128]: Do you have a production use case? 

I am supportive of making these available to e2e test though.

File: tests/linearizability/failpoints.go [hunk 11]
```diff
@@ -99,13 +103,92 @@ func (f killFailpoint) Trigger(t *testing.T, ctx context.Context, clus *e2e.Etcd
...
+	endpoints := make([]string, 0, len(clus.EndpointsV3()))
+	for _, ed := range clus.EndpointsV3() {
+		if ed != member.EndpointsV3()[0] {
+			endpoints = append(endpoints, ed)
+		}
+	}
+
+	if f.waitForSnapshot {
+		// wait for potential leader election
+		time.Sleep(2 * time.Second)
```
[the remark this answers is not in the document]
Comment (chaochn47) [reply tests/linearizability/failpoints.go:116]: Thanks for the call out.

File: tests/linearizability/failpoints.go (hunk 8 above)
Comment (chaochn47) [reply tests/linearizability/failpoints.go:130]: Yeah, line 134 is restart the member.

File: tests/linearizability/failpoints.go (hunk 7 above)
Comment (chaochn47) [reply tests/linearizability/failpoints.go:163]: ack

**Comment by chaochn47** [comment]
Looks like in the CI box particularlly, 

```
logger.go:130: 2022-12-20T01:44:47.062Z	INFO	Issue14571	linearizability/failpoints.go:169	got revision	{"initial-revision": 379, "current-revision": 379, "snapshot-catchup-count": 5000}
```

The write does not go though. Need to investigate more before publishing it to review.

File: tests/linearizability/failpoints.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (lavacat) [reply tests/linearizability/failpoints.go:128]: I don't have prod case and the comment in config.go says
>// WARNING: only change this for tests. Always use "DefaultSnapshotCatchUpEntries"
On one hand, it's not a good idea to expose flag just because of testing, but doing 5k ops to trigger snapshot apply is also not great.

Another e2e test use case:
https://github.com/etcd-io/etcd/pull/15013#discussion_r1051683699

[the remark this answers is not in the document]
Comment (ahrtr) [reply tests/linearizability/failpoints.go:128]: I agree to make `DefaultSnapshotCatchUpEntries` configuable via a user-facing flag, two reasons:
1. It can benefit edge case, which usually is one-node cluster, and with limited resource (e.g memory). Setting a small value for `DefaultSnapshotCatchUpEntries` can reduce the memory usage. 
2. It can benefit the E2E test case.

Please let me know if any of you can do this, otherwise let me take care of it.

[the remark this answers is not in the document]
Comment (chaochn47) [reply tests/linearizability/failpoints.go:128]: Thanks @ahrtr I can take that as it looks like blocking my current PR. 

Because I just realize that linearizability test will repeat 60 times.. The current PR will add average 1min execution time per run. So It's not acceptable.  

https://github.com/etcd-io/etcd/blob/a59276c171d1647653b75679a12f85068d3b192b/.github/workflows/linearizability.yaml#L17

File: tests/linearizability/failpoints.go (hunk 9 above)
Comment (chaochn47) [reply tests/linearizability/failpoints.go:91]: Second thought, we can definitely optimize that with your suggestion after stabilizing this test case..

File: tests/linearizability/failpoints.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (chaochn47) [reply tests/linearizability/failpoints.go:128]: ack. Will do on PST daytime and cc you and other stakeholders.

**Comment by chaochn47** [comment]
Tested with v3.5.5 binary with experimental-snapshot-catchup-entry commit. `linearizability_test.go:277: Model is not linearizable`

**Comment by ahrtr** [comment]
> Tested 60 times with v3.5.5 binary with experimental-snapshot-catchup-entry commit. `linearizability_test.go:277: Model is not linearizable` while on mainline, it is.

Thanks @chaochn47, please provide a detail steps. I may take a look sometime on weekend or next week.

**Comment by chaochn47** [comment]
Hi @ahrtr, it is not a new issue. The test is added to reproduce and prevent #14571. It has been fixed in v3.5.6.

**Comment by chaochn47** [comment]
The PR is ready for review.

@ptabor @serathius Could you please take a second look, thanks!

**Comment by serathius** [comment]
Please  resolve conficts.

**Comment by chaochn47** [comment]
```
TestLinearizability_ClusterOfSize3

linearizability_test.go:428: Linearization timed out
```

**Comment by chaochn47** [comment]
Conflicts have been resolved. Would you mind take a second look? @serathius Thanks!!

File: tests/linearizability/failpoints.go [hunk 12]
```diff
@@ -33,7 +36,8 @@ const (
 )
 
 var (
-	KillFailpoint                            Failpoint = killFailpoint{}
+	KillFailpoint                            Failpoint = killFailpoint{target: AnyMember}
+	EnableAuthKillFailpoint                  Failpoint = killFailpoint{enableAuth: true, target: Follower}
```
Comment (serathius) [inline tests/linearizability/failpoints.go:40]: Do you need to target follower here? I understand that 1471 happens when follower is killed, however we don't need to hardcode it.

Comment (serathius) [inline tests/linearizability/failpoints.go:40]: I don't understand why failpoint needs to be aware of auth. It's true that at this moment it creates the client, but maybe we can move client creation to different place. Either do dependency injection and provide failpoint with client that is already authorized. Alternative would be to add method to e2e.EtcdProcessCluster that provides the client.

File: tests/linearizability/failpoints.go [hunk 13]
```diff
@@ -101,11 +112,19 @@ func (f killFailpoint) Trigger(t *testing.T, ctx context.Context, clus *e2e.Etcd
...
+	endpoints := make([]string, 0, len(clus.EndpointsV3()))
+	for _, ed := range clus.EndpointsV3() {
+		if ed != member.EndpointsV3()[0] {
+			endpoints = append(endpoints, ed)
+		}
 	}
-	return nil
+
+	if f.enableAuth {
+		require.NoError(t, addTestUserAuth(ctx, endpoints))
```
Comment (serathius) [inline tests/linearizability/failpoints.go:124]: Don't like that auth setup is part of failpoint injection. Those are totally separate things. Please move auth setup to cluster setup.

File: tests/linearizability/linearizability_test.go [hunk 14]
```diff
@@ -54,15 +58,22 @@ var (
...
 	tcs := []struct {
-		name      string
-		failpoint Failpoint
-		config    e2e.EtcdProcessClusterConfig
-		traffic   *trafficConfig
+		name        string
+		failpoint   Failpoint
+		config      e2e.EtcdProcessClusterConfig
+		traffic     *trafficConfig
+		clientCount int
```
Comment (serathius) [inline tests/linearizability/linearizability_test.go:76]: This client count doesn't seem to be used. Please remove it and use `TrafficConfig.clientCount`

File: tests/linearizability/linearizability_test.go [hunk 15]
```diff
@@ -124,39 +149,40 @@ func TestLinearizability(t *testing.T) {
 				t.Fatal(err)
 			}
 			defer clus.Close()
+			lg := zaptest.NewLogger(t, zaptest.WrapOptions(zap.AddCaller())).Named(tc.name)
```
Comment (serathius) [inline tests/linearizability/linearizability_test.go:152]: This is nice, but please consider moving it to separate PR.

File: tests/linearizability/linearizability_test.go [hunk 16]
```diff
@@ -277,21 +303,26 @@ type FailpointConfig struct {
...
 	lm := identity.NewLeaseIdStorage()
-	h := model.History{}
+	h := &model.History{}
 	limiter := rate.NewLimiter(rate.Limit(config.maximalQPS), 200)
 
 	startTime := time.Now()
 	wg := sync.WaitGroup{}
+
 	for i := 0; i < config.clientCount; i++ {
+		i := i
```
Comment (serathius) [inline tests/linearizability/linearizability_test.go:321]: This should not be needed as we pass `i` as argument to goroutine funtion.

File: tests/linearizability/linearizability_test.go [hunk 17]
```diff
@@ -301,23 +332,62 @@ func simulateTraffic(ctx context.Context, t *testing.T, clus *e2e.EtcdProcessClu
...
 	qps := float64(len(operations)) / float64(endTime.Sub(startTime)) * float64(time.Second)
-	t.Logf("Average traffic: %f qps", qps)
+	lg.Info("Average traffic", zap.Float64("qps", qps))
 	if qps < config.minimalQPS {
 		t.Errorf("Requiring minimal %f qps for test results to be reliable, got %f qps", config.minimalQPS, qps)
 	}
 	return operations
 }
 
+func simulatePostFailpointTraffic(ctx context.Context, wg *sync.WaitGroup, endpoints []string, clientId int, ids identity.Provider, h *model.History, mux *sync.Mutex, config trafficConfig, limiter *rate.Limiter, lm identity.LeaseIdStorage) {
```
Comment (serathius) [inline tests/linearizability/linearizability_test.go:355]: Don't understand why you cannot incorporate this into normal traffic

File: tests/linearizability/traffic.go [hunk 18]
```diff
@@ -17,39 +17,57 @@ package linearizability
...
 type opChance struct {
 	operation model.OperationType
 	chance    int
 }
 
-func (t readWriteSingleKey) Run(ctx context.Context, clientId int, c *recordingClient, limiter *rate.Limiter, ids identity.Provider, lm identity.LeaseIdStorage) {
+func (t readWriteSingleKey) PreRun(ctx context.Context, c interfaces.Client, lg *zap.Logger) error {
+	if t.AuthEnabled() {
+		lg.Info("set up auth")
+		return setupAuth(ctx, c)
```
Comment (serathius) [inline tests/linearizability/traffic.go:65]: Authorization setup should be done at cluster setup, not at this point.

File: tests/linearizability/linearizability_test.go (hunk 17 above)
Comment (chaochn47) [reply tests/linearizability/linearizability_test.go:355]: To incorporate this into normal traffic, a client with test user authorization has to be set up upfront. However, before the user is added to the cluster, client creation will fail. 

After a couple of failed attempts, I adopted this workaround. It's more deterministic to reliably reproduce on the impacted 3.5 versions.

File: tests/linearizability/failpoints.go (hunk 12 above)
Comment (chaochn47) [reply tests/linearizability/failpoints.go:40]: Good suggestion! I can explore each option and see what's the best fit.

File: tests/linearizability/traffic.go (hunk 18 above)
Comment (chaochn47) [reply tests/linearizability/traffic.go:65]: Make sense.

File: tests/linearizability/failpoints.go (hunk 13 above)
Comment (serathius) [reply tests/linearizability/failpoints.go:124]: I think that enabling auth is orthogonal to failure injection.

File: tests/linearizability/linearizability_test.go (hunk 17 above)
Comment (serathius) [reply tests/linearizability/linearizability_test.go:355]: Please setup authorization in cluster setup.

**Comment by chaochn47** [comment]
Revisit this task again, I think the linearizability test does not have to reproduce the exact same scenario how #14571 happened. 

#14571 uncovers the issue that auth recover from snapshot failed to update `rangePermCache`. It's a in-memory map, for each user, maintains a interval tree that checks requested key start to key end is the subnet of the interval tree. 

The plan will be

At cluster setup stage: 
1. create root user, root role has access to all operations. 
2. create test user, test role, grant test role RW permissions from key `foo` to `zoo`
Traffic generator 
1. half of the clients using root user permissions
2. half of the clients using test user permissions
3. root user client will periodically grant / revoke test role permissions

Fault injector
1. kill a random member
